cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java

```diff
@@ -228,16 +238,64 @@ public static CruiseControl fromCrd(Kafka kafkaAssembly, KafkaVersion.Lookup ver
-        for (Map.Entry<String, String> entry : configuration.getCruiseControlDefaultPropertiesMap().entrySet()) {
-            if (configuration.getConfigOption(entry.getKey()) == null) {
-                configuration.setConfigOption(entry.getKey(), entry.getValue());
+        CruiseControlConfiguration userConfiguration = new CruiseControlConfiguration(spec.getConfig().entrySet());
+        for (Map.Entry<String, String> defaultEntry : userConfiguration.getCruiseControlDefaultPropertiesMap().entrySet()) {
+            if (userConfiguration.getConfigOption(defaultEntry.getKey()) == null) {
+                userConfiguration.setConfigOption(defaultEntry.getKey(), defaultEntry.getValue());
             }
         }
-        cruiseControl.setConfiguration(configuration);
+        // Ensure that the configured anomaly.detection.goals are a sub-set of the default goals
+        checkGoals(userConfiguration);
+        cruiseControl.setConfiguration(userConfiguration);
         return cruiseControl;
     }
 
+    /**
+     *  This method ensures that the checks in cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaCruiseControlConfig.java
+     *  sanityCheckGoalNames() method (L118)  don't fail if a user submits custom default goals that have less members then the default
+     *  anomaly.detection.goals.
+     * @param configuration The configuration instance to be checked.
+     * @throws UnsupportedOperationException If the configuration contains self.healing.goals configurations.
+     */
+    public static void checkGoals(CruiseControlConfiguration configuration) {
+        // If self healing goals are defined then these take precedence
+        if (configuration.getConfigOption(CRUISE_CONTROL_SELF_HEALING_CONFIG_KEY) != null) {
+            String selfHealingGoalsString = configuration.getConfigOption(CRUISE_CONTROL_ANOMALY_DETECTION_CONFIG_KEY);
+            List<String> selfHealingGoals = Arrays.asList(selfHealingGoalsString.split("\\s*,\\s*"));
+            if (!selfHealingGoals.isEmpty()) {
+                throw new UnsupportedOperationException("Cruise Control's self healing functionality is not currently supported. Please remove " +
+                        CRUISE_CONTROL_SELF_HEALING_CONFIG_KEY + " config");
+            }
+        }
+
+        // Either self.healing.goals is null or it is an empty list
+        String anomalyGoalsString;
+        if (configuration.getConfigOption(CRUISE_CONTROL_ANOMALY_DETECTION_CONFIG_KEY) != null) {
+            // The user has defined some anomaly detection goals
+            anomalyGoalsString = configuration.getConfigOption(CRUISE_CONTROL_ANOMALY_DETECTION_CONFIG_KEY);
+        } else {
+            // No anomaly detection goals have been defined by the user, therefore the defaults defined in Cruise Control
+            // will be used.
+            anomalyGoalsString = CRUISE_CONTROL_DEFAULT_ANOMALY_DETECTION_GOALS;
+        }
```

Comment:
If no `anomaly.detection.goals` are specified by the user, I believe the CC defaults[1] are set by CC automatically. This else-statement may not be needed. 

However, I agree we need to make sure that the `anomaly.detection.goals` need to be a subset of `goals`[2] list we provide, whether we add missing goals from the default `anomaly.detection.goals`[1] to the supported goal list [2] or declare a modified version of the default `anomaly.detection.goals` here [3]

[1] Default `anomaly.detection.goals`
```com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareCapacityGoal, com.linkedin.kafka.cruisecontrol.analyzer.goals.PotentialNwOutGoal, com.linkedin.kafka.cruisecontrol.analyzer.goals.ResourceDistributionGoal, com.linkedin.kafka.cruisecontrol.analyzer.goals.LeaderBytesInDistributionGoals, com.linkedin.kafka.cruisecontrol.analyzer.goals.TopicReplicaDistributionGoal, com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal```
[2] https://github.com/tomncooper/strimzi-kafka-operator/blob/b4c02a4a3aa2d1a65049d21039616f6cd67a10d1/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java#L26-L41
[3] https://github.com/tomncooper/strimzi-kafka-operator/blob/b4c02a4a3aa2d1a65049d21039616f6cd67a10d1/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControlConfiguration.java#L60

Reply:
I need to know what the anomaly detection goals are (wether set by the user or default from CC) in order to check weather they are a subset of the default.goals. So that is why that If statement if there. ADGs need to be a subset of **default.goals** specifically as these are what are checked in the CC startup code.

Where did you get the default ADG list from? I am taking the one from L104 of `cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnomalyDetectorConfig.java` which only lists 3: RackAwareness, ReplicaCapacity and DiskCapacity.

Reply:
Ah I understand now!

> Where did you get the default ADG list from?

I was getting it from the `anomaly.detectio.goals` list from the CC Configuration documentation [1] but it appears to be outdated! You are right, the source code is the source of truth!

[1] https://github.com/linkedin/cruise-control/wiki/Configurations